src/Draco.Compiler/Internal/Syntax/ParseTreeFormatter.cs

```diff
@@ -0,0 +1,179 @@
+using System.Text;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Draco.Compiler.Api.Syntax;
+using Token = Draco.Compiler.Internal.Syntax.ParseTree.Token;
+
+namespace Draco.Compiler.Internal.Syntax;
+
+internal sealed record class ParseTreeFormatterSettings(string Indentation)
+{
+    internal static readonly ParseTreeFormatterSettings DefaultSettings = new ParseTreeFormatterSettings("    ");
+}
+
+internal sealed class ParseTreeFormatter : ParseTreeTransformerBase
+{
+    private TokenType? lastToken;
+    private TokenType? nextToken;
+    private IEnumerator<Token>? tokens;
+    private readonly ParseTreeFormatterSettings settings;
+    private int indentCount = 0;
+    private string Indentation
+    {
+        get
+        {
+            var result = new StringBuilder();
+            for (int i = 0; i < this.indentCount; ++i) result.Append(this.settings.Indentation);
+            return result.ToString();
+        }
+    }
+
+    internal ParseTreeFormatter(ParseTreeFormatterSettings settings)
+    {
+        this.settings = settings;
+    }
+
+    private IEnumerable<Token> GetTokens(ParseTree tree) =>
+        tree.InOrderTraverse().OfType<Token>();
+
+    private Token.Builder AddIndentation(Token.Builder newToken)
+    {
+        this.indentCount++;
+        return newToken;
+    }
+
+    private Token.Builder RemoveIndentation(Token.Builder newToken)
+    {
+        this.indentCount--;
+        return newToken;
+    }
+
+    public ParseTree Format(ParseTree tree)
+    {
+        this.tokens = this.GetTokens(tree).GetEnumerator();
+        // We need to be one token ahead, because the next token affects the current one, so we must advance twice here
+        if (!(this.tokens.MoveNext() && this.tokens.MoveNext()))
+            return tree;
+        this.nextToken = this.tokens.Current.Type;
+        return this.Transform(tree, out _);
+    }
+
+    public override Token TransformToken(Token token, out bool changed)
+    {
+        if (token.Type == TokenType.EndOfInput)
+        {
+            changed = false;
+            return token;
+        }
+        var newToken = Token.Builder.From(token);
+
+        newToken = newToken.Type switch
+        {
+            TokenType.Assign or TokenType.Colon or TokenType.Comma or TokenType.Equal or
+            TokenType.GreaterEqual or TokenType.GreaterThan or TokenType.InterpolationStart or
+            TokenType.KeywordAnd or TokenType.KeywordFrom or TokenType.KeywordImport or
+            TokenType.KeywordMod or TokenType.KeywordNot or TokenType.KeywordOr or
+            TokenType.KeywordRem or TokenType.LessEqual or TokenType.LessThan or
+            TokenType.Minus or TokenType.MinusAssign or TokenType.NotEqual or
+            TokenType.Plus or TokenType.PlusAssign or TokenType.Slash or
+            TokenType.SlashAssign or TokenType.Star or TokenType.StarAssign
+            => newToken.SetTrailingTrivia(this.CreateTrivia(TokenType.Whitespace, " ")),
```

Comment:
You use `this.CreateTrivia(TokenType.Whitespace, " ")`, it might be constant-worthy!